File: csharp/src/EnvoiceClient.cs
using System.Net.Http.Json;
using System.Text.Json;

namespace Envoice;

/// <summary>Main client for interacting with the envoice.dev API</summary>
public class EnvoiceClient : IDisposable
{
    private readonly string _apiKey;
    private readonly string _apiUrl;
    private readonly HttpClient _httpClient;
    private readonly JsonSerializerOptions _jsonOptions;
    private readonly bool _ownsHttpClient;

    /// <summary>Create a new EnvoiceClient</summary>
    public EnvoiceClient(string apiKey, string? apiUrl = null, TimeSpan? timeout = null, HttpClient? httpClient = null)
    {
        if (string.IsNullOrWhiteSpace(apiKey))
            throw new ArgumentException("API key is required", nameof(apiKey));

        _apiKey = apiKey;
        _apiUrl = (apiUrl ?? "https://api.envoice.dev").TrimEnd('/');
        _ownsHttpClient = httpClient == null;
        _httpClient = httpClient ?? new HttpClient { Timeout = timeout ?? TimeSpan.FromSeconds(30) };
        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
        };
    }

    /// <summary>Create a new invoice builder</summary>
    public InvoiceBuilder Invoice() => new(this);

    /// <summary>Generate an invoice directly</summary>
    public async Task<InvoiceResult> GenerateInvoiceAsync(GenerateRequest request, CancellationToken cancellationToken = default)
    {
        try
        {
            using var httpRequest = new HttpRequestMessage(HttpMethod.Post, $"{_apiUrl}/v1/generate")
            {
                Content = JsonContent.Create(request, options: _jsonOptions)
            };
            httpRequest.Headers.Add("X-API-Key", _apiKey);

            using var response = await _httpClient.SendAsync(httpRequest, cancellationToken);

            return await HandleGenerateResponseAsync(response, cancellationToken);
        }
        catch (HttpRequestException ex)
        {
            throw new EnvoiceNetworkException("Network error", ex);
        }
        catch (TaskCanceledException ex) when (ex.InnerException is TimeoutException)
        {
            throw new EnvoiceNetworkException("Request timeout", ex);
        }
    }

    private async Task<InvoiceResult> HandleGenerateResponseAsync(HttpResponseMessage response, CancellationToken cancellationToken)
    {
        var content = await response.Content.ReadAsStringAsync(cancellationToken);

        if (response.IsSuccessStatusCode)
        {
            var data = JsonSerializer.Deserialize<GenerateResponse>(content, _jsonOptions)!;
            return new InvoiceSuccess(data.PdfBase64, data.Filename, data.Validation, data.Account);
        }

        ErrorResponse? error;
        try
        {
            error = JsonSerializer.Deserialize<ErrorResponse>(content, _jsonOptions);
        }
        catch
        {
            error = new ErrorResponse("unknown_error", $"HTTP {(int)response.StatusCode}");
        }

        var statusCode = (int)response.StatusCode;

        if (statusCode == 402)
        {
            throw new EnvoiceQuotaExceededException(error?.Message ?? "Quota exceeded");
        }

        if (statusCode == 422 && error?.Details != null)
        {
            return new InvoiceFailure(error.Details);
        }

        throw new EnvoiceApiException(
            error?.Message ?? error?.Error ?? "Unknown error",
            statusCode,
            error?.Error
        );
    }

    /// <summary>Validate an existing PDF</summary>
    public async Task<ValidationResult> ValidateAsync(string pdfBase64, CancellationToken cancellationToken = default)
    {
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, $"{_apiUrl}/v1/validate")
            {
                Content = JsonContent.Create(new { pdf_base64 = pdfBase64 })
            };
            request.Headers.Add("X-API-Key", _apiKey);

            using var response = await _httpClient.SendAsync(request, cancellationToken);

            if (!response.IsSuccessStatusCode)
            {
                var error = await response.Content.ReadFromJsonAsync<ErrorResponse>(cancellationToken: cancellationToken);
                throw new EnvoiceApiException(
                    error?.Message ?? "Validation failed",
                    (int)response.StatusCode,
                    error?.Error
                );
            }

            return (await response.Content.ReadFromJsonAsync<ValidationResult>(cancellationToken: cancellationToken))!;
        }
        catch (HttpRequestException ex)
        {
            throw new EnvoiceNetworkException("Network error", ex);
        }
    }

    /// <summary>Get account information</summary>
    public async Task<AccountInfo> GetAccountAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, $"{_apiUrl}/v1/account");
            request.Headers.Add("X-API-Key", _apiKey);

            using var response = await _httpClient.SendAsync(request, cancellationToken);

            if (!response.IsSuccessStatusCode)
            {
                var error = await response.Content.ReadFromJsonAsync<ErrorResponse>(cancellationToken: cancellationToken);
                throw new EnvoiceApiException(
                    error?.Message ?? "Failed to get account",
                    (int)response.StatusCode,
                    error?.Error
                );
            }

            return (await response.Content.ReadFromJsonAsync<AccountInfo>(cancellationToken: cancellationToken))!;
        }
        catch (HttpRequestException ex)
        {
            throw new EnvoiceNetworkException("Network error", ex);
        }
    }

    public void Dispose()
    {
        if (_ownsHttpClient)
        {
            _httpClient.Dispose();
        }
        GC.SuppressFinalize(this);
    }
}

File: java/src/main/java/dev/envoice/EnvoiceClient.java
package dev.envoice;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

/**
 * Main client for interacting with the envoice.dev API
 */
public final class EnvoiceClient implements AutoCloseable {
    private final String apiKey;
    private final String apiUrl;
    private final Duration timeout;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    /**
     * Create a new EnvoiceClient with default options
     */
    public EnvoiceClient(String apiKey) {
        this(apiKey, "https://api.envoice.dev", Duration.ofSeconds(30));
    }

    /**
     * Create a new EnvoiceClient with custom options
     */
    public EnvoiceClient(String apiKey, String apiUrl, Duration timeout) {
        if (apiKey == null || apiKey.isBlank()) {
            throw new IllegalArgumentException("API key is required");
        }
        this.apiKey = apiKey;
        this.apiUrl = apiUrl.replaceAll("/$", "");
        this.timeout = timeout;
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();
        this.objectMapper = new ObjectMapper()
            .registerModule(new JavaTimeModule());
    }

    /**
     * Create a new invoice builder
     */
    public InvoiceBuilder invoice() {
        return new InvoiceBuilder(this);
    }

    /**
     * Generate an invoice directly
     */
    public InvoiceResult generateInvoice(Types.GenerateRequest request) {
        try {
            String body = objectMapper.writeValueAsString(request);

            HttpRequest httpRequest = HttpRequest.newBuilder()
                .uri(URI.create(apiUrl + "/v1/generate"))
                .header("Content-Type", "application/json")
                .header("X-API-Key", apiKey)
                .timeout(timeout)
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();

            HttpResponse<String> response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());

            return handleGenerateResponse(response);
        } catch (IOException e) {
            throw new Exceptions.EnvoiceNetworkException("Network error", e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new Exceptions.EnvoiceNetworkException("Request interrupted", e);
        }
    }

    private InvoiceResult handleGenerateResponse(HttpResponse<String> response) throws IOException {
        int status = response.statusCode();
        String body = response.body();

        if (status == 200) {
            Types.GenerateResponse data = objectMapper.readValue(body, Types.GenerateResponse.class);
            return new InvoiceResult.Success(
                data.pdfBase64(),
                data.filename(),
                data.validation(),
                data.account()
            );
        }

        Types.ErrorResponse error;
        try {
            error = objectMapper.readValue(body, Types.ErrorResponse.class);
        } catch (Exception e) {
            error = new Types.ErrorResponse("unknown_error", "HTTP " + status, null);
        }

        if (status == 402) {
            throw new Exceptions.EnvoiceQuotaExceededException(
                error.message() != null ? error.message() : "Quota exceeded"
            );
        }

        if (status == 422 && error.details() != null) {
            return new InvoiceResult.Failure(error.details());
        }

        throw new Exceptions.EnvoiceApiException(
            error.message() != null ? error.message() : error.error(),
            status,
            error.error()
        );
    }

    /**
     * Validate an existing PDF
     */
    public Types.ValidationResult validate(String pdfBase64) {
        try {
            String body = objectMapper.writeValueAsString(java.util.Map.of("pdf_base64", pdfBase64));

            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiUrl + "/v1/validate"))
                .header("Content-Type", "application/json")
                .header("X-API-Key", apiKey)
                .timeout(timeout)
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                Types.ErrorResponse error = objectMapper.readValue(response.body(), Types.ErrorResponse.class);
                throw new Exceptions.EnvoiceApiException(
                    error.message() != null ? error.message() : error.error(),
                    response.statusCode(),
                    error.error()
                );
            }

            return objectMapper.readValue(response.body(), Types.ValidationResult.class);
        } catch (IOException e) {
            throw new Exceptions.EnvoiceNetworkException("Network error", e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new Exceptions.EnvoiceNetworkException("Request interrupted", e);
        }
    }

    /**
     * Get account information
     */
    public Types.AccountInfo getAccount() {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiUrl + "/v1/account"))
                .header("X-API-Key", apiKey)
                .timeout(timeout)
                .GET()
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                Types.ErrorResponse error = objectMapper.readValue(response.body(), Types.ErrorResponse.class);
                throw new Exceptions.EnvoiceApiException(
                    error.message() != null ? error.message() : error.error(),
                    response.statusCode(),
                    error.error()
                );
            }

            return objectMapper.readValue(response.body(), Types.AccountInfo.class);
        } catch (IOException e) {
            throw new Exceptions.EnvoiceNetworkException("Network error", e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new Exceptions.EnvoiceNetworkException("Request interrupted", e);
        }
    }

    @Override
    public void close() {
        // HttpClient doesn't need explicit closing in this implementation
    }
}
